packages/lab/src/dropdown-next/DropdownBase.tsx

```diff
@@ -0,0 +1,182 @@
+import { flip, limitShift, shift, size } from "@floating-ui/react";
+import {
+  makePrefixer,
+  useFloatingUI,
+  useForkRef,
+  useIdMemo as useId,
+} from "@salt-ds/core";
+import { clsx } from "clsx";
+import { Children, cloneElement, forwardRef, useRef, useState } from "react";
+import { forwardCallbackProps } from "../utils";
+import { DropdownBaseProps } from "./dropdownTypes";
+import { useDropdownBase } from "./useDropdownBase";
+import { Portal } from "../portal";
+import { isDesktop, useWindow } from "../window";
+
+import "./DropdownNext.css";
+
+// Any component may be passed as our trigger or popup component.
+// Define the common props that we will act on, if present,
+// so we can type them.
+export type MaybeChildProps = {
+  className?: string;
+  id?: string;
+  role?: string;
+  width: number | string;
+};
+
+const withBaseName = makePrefixer("saltDropdown");
+
+export const DropdownBase = forwardRef<HTMLDivElement, DropdownBaseProps>(
+  function Dropdown(
+    {
+      "aria-labelledby": ariaLabelledByProp,
+      children,
+      className: classNameProp,
+      container,
+      defaultIsOpen,
+      disabled,
+      disablePortal,
+      fullWidth,
+      id: idProp,
+      isOpen: isOpenProp,
+      onKeyDown,
+      onOpenChange,
+      openOnFocus,
+      placement = "bottom-start",
+      popupWidth,
+      width,
+      variant = "primary",
+      ...htmlAttributes
+    },
+    forwardedRef
+  ) {
+    const rootRef = useRef<HTMLDivElement>(null);
+    const className = clsx(withBaseName(), classNameProp, {
+      [withBaseName("fullWidth")]: fullWidth,
+      [withBaseName("disabled")]: disabled,
+    });
+    const [trigger, popupComponent] = Children.toArray(
+      children
+    ) as JSX.Element[];
+    const id = useId(idProp);
+    const Window = useWindow();
+
+    const { componentProps, popperRef, isOpen, triggerProps } = useDropdownBase(
+      {
+        ariaLabelledBy: ariaLabelledByProp,
+        defaultIsOpen,
+        disabled,
+        fullWidth,
+        id,
+        isOpen: isOpenProp,
+        onOpenChange,
+        onKeyDown,
+        openOnFocus,
+        popupComponent,
+        popupWidth,
+        rootRef,
+        width,
+      }
+    );
+    const [maxPopupHeight, setMaxPopupHeight] = useState<number | undefined>(
+      undefined
+    );
+
+    const middleware = isDesktop
```

Comment:
the isDesktop boolean will likely be removed in favor of a cleaner solution.

Perhaps we could remove it early.